### PR TITLE
WIP: Visualizer

### PR DESCRIPTION
**Pull request by ivrolan**
Hi,

Please take a look to the initial draft Visualizer class in `_visualizer_tools.py`. This visualizer iterates over a list `elements` of `AbstractDrawer` that draw the required object with the function `draw_at_t(t)`. I included the time as a parameter as I thought the properties of the object to draw are functions of time, so we need the specific point of time to draw it.

 Each child class of `AbstractDrawer` should contain the object to draw and additional properties to draw it, i.e. the color of the ball. 

However, I haven't been able to run the test because of type conflicts, still working on it. What do you think of this approach?

### CODE REVIEW

**Comment by knyazer** [comment]
ping me directly if you want me to take a look, otherwise i don't see it: either do @knyazer, or request me as a reviewer

give me a sec to take a look

**Comment by knyazer** [comment]
Overall quite a decent work. However, there are few suggestions about design:

firstly, we can make a drawer that would work independent of a particular body: any body must provide ```shape: UniversalShape``` and any ```UniversalShape``` have a bunch of ```parts``` that are ```AbstractConvexShape```s, from which you can use ```get_support``` to figure out a point on the surface of the shape. Then you can draw anything via simply connecting a bunch of points on the surface of every convex shape. I agree that it requires a bit more work, but this also reduce the connectedness of the source code, hence it is much easier to support: once written, we can zero-shot support any shape

**Comment by knyazer** [comment]
But I am not going to accept a PR that fails tests: for obvious reasons, failing tests mean that something is clearly wrong.

**Comment by knyazer** [comment]
And somehow even after pre-commit hook you have issues with formatting: are you sure that you have followed ```CONTRIBUTING.md```?

**Comment by knyazer** [comment]
> `draw_at_t(t)`. I included the time as a parameter as I thought the properties of the object to draw are functions of time, so we need the specific point of time to draw it.

I don't precisely understand what you mean: while the thing we draw is indeed a function of time, we never have a dependency on it. We simply plot the current state of the system while it evolves through time.

**Comment by ivrolan** [comment]
> > `draw_at_t(t)`. I included the time as a parameter as I thought the properties of the object to draw are functions of time, so we need the specific point of time to draw it.
> 
> I don't precisely understand what you mean: while the thing we draw is indeed a function of time, we never have a dependency on it. We simply plot the current state of the system while it evolves through time.

So can we take the values of position, velocity, etc as they are in the object? Will these values be changed in other places of the code to simulate in continous time?


> And somehow even after pre-commit hook you have issues with formatting: are you sure that you have followed `CONTRIBUTING.md`?

Fixed! I was not using `pre-commit` as it was intended




> But I am not going to accept a PR that fails tests: for obvious reasons, failing tests mean that something is clearly wrong.

I was not expecting otherwise, that is why this is marked as a Draft for now :)


Thanks for the feedback! I'll remove the `t` parameter of the drawing function and tweak the code to make it pass the tests first

**Comment by ivrolan** [comment]
> Overall quite a decent work. However, there are few suggestions about design:
> 
> firstly, we can make a drawer that would work independent of a particular body: any body must provide `shape: UniversalShape` and any `UniversalShape` have a bunch of `parts` that are `AbstractConvexShape`s, from which you can use `get_support` to figure out a point on the surface of the shape. Then you can draw anything via simply connecting a bunch of points on the surface of every convex shape. I agree that it requires a bit more work, but this also reduce the connectedness of the source code, hence it is much easier to support: once written, we can zero-shot support any shape

Even if we can make an independent drawer based on the shape, I think we can also have more properties to take into account when drawing than the shape of the object. For example, for a robot, we can add a number id on top of it to visualize the agent or a change of color when driving with the goal.

Basically, it would be useful to have a general shape drawing tool, but we will need to add more things on top of it depending on the object to draw.

**Comment by knyazer** [comment]
Almost LGTM: can you please replace BallDrawer with some sort of Drawer that would work on any AbstractBody? Sample a lot of point in the circle for every convex component of abstract body shape, and draw these things according to the body transformation matrix. Feel free to introduce a getter for such a transformation matrix into AbstractBody

File: cotix/_visualizer_tools.py [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+        # retrieve the first coord of the global support as it should be equal
+        # to the radius
+        # for more complex shapes we will need more info (get 360 points around
+        # a circle with the global_support?)
+        radius = self.object.shape.get_global_support(jnp.array([1.0, 0.0]))[0]
+
+        # convert meters to pixels with scale
+        center_x *= pix2meter
+        center_y *= pix2meter
+        radius *= pix2meter
```
Comment (knyazer) [inline cotix/_visualizer_tools.py:43]: ins't meter to pix a better name? Since you want to convert meters to pixels

File: test/test_viz.py [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+        # Iterating over all the events received from
+        # pygame.event.get()
+        for event in viz.get_events():
+            # If the type of the event is quit
+            # then setting the run variable to false
+            if event.type == pygame.locals.QUIT:
+                run = False
+        # Draws the surface object to the screen.
+
+        viz.draw([b_drawer])
```
Comment (knyazer) [inline test/test_viz.py:36]: Why do you have to draw b_drawer? Just draw all the objects passed. Like,
```viz.draw([Ball(), Robot(), ...])``` and viz draw could look something like
```
for x in objects:
    draw_body(x)
```

File: test/test_viz.py [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+
+        framecount += 1
+
+        # update the ball
+        b = b.set_position(b.position + inc)
+        # update the drawer with the ball
+        b_drawer.object = b
+
+        if framecount == framerate:
+            run = False
```
Comment (knyazer) [inline test/test_viz.py:48]: Seems like a bad way to write condition. why exatcly this value? 

And, besides, you still don't make any assertions in the test: it just checks that you have pygame installed. Maybe you can think about some meaningful property that you can test, like, that you actually draw something? Or just remove the test completely.

If you want to test that some functionality works, like that it draws something, don't commit these files: these tests are meaningless for other people, so they should not be shared

File: test/test_viz.py [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+        viz.draw([b_drawer])
+
+        pygame.time.Clock().tick(framerate)
+
+        framecount += 1
+
+        # update the ball
+        b = b.set_position(b.position + inc)
+        # update the drawer with the ball
+        b_drawer.object = b
```
Comment (knyazer) [inline test/test_viz.py:45]: Also you are not allowed to do b_drawer.object = b

Comment (knyazer) [reply test/test_viz.py:45]: I mean, since it is not JAX, you are allowed to do so, but since we are trying to make your simulator in the same style as JAX, it should adhere

BTW, one cool test that it works under jit would be that you use ```python jax.debug.callback``` to do any pygame-related stuff, and then just put everything you do into jitted region. Then you will be able to see whether your code actually adheres to the standard (since otherwise it won't compile), and it is also more convenient for us, since then your simulator will be less restrictive in terms of control flow: now we could do calls from inside jit, if we want to, and allow the user decide when to do visualizations. I'd like to see this implemented, but it might be a bit too complex

Comment (ivrolan) [reply test/test_viz.py:45]: So instead of doing 

       b_drawer.object = b

We should create the b_drawer from scratch with the new b info, right?

File: test/test_viz.py (hunk 2 above)
Comment (ivrolan) [reply test/test_viz.py:36]: Because the information in the objects `Ball(), Robot()` is not enough to draw them, e.g. the color. We can add this data as properties of the objects, but as this is only visualization related, I think it is better to create other classes called `BallDrawer(), RobotDrawer()` that draw the corresponding object according to additional data passed

File: cotix/_visualizer_tools.py (hunk 1 above)
Comment (ivrolan) [reply cotix/_visualizer_tools.py:43]: True, my bad! I was thinking about `pix_by_meter` but I end up writing other thing

File: test/test_viz.py (hunk 3 above)
Comment (ivrolan) [reply test/test_viz.py:48]: As I want the simulation to run over 1s, the number of frames executed will be equal to the framerate you're achieving.

Thinking about a way to test with assertions, we could run the viz, save the image and then comparing future tests simulations to this image. In that way, we can test that everything is drawn in the same way as our "correct" run

File: test/test_viz.py (hunk 4 above)
Comment (knyazer) [reply test/test_viz.py:45]: either this, or better you could use ```eqx.tree_at(lambda drawer: drawer.object, b_drawer, replace=b)```

File: test/test_viz.py (hunk 3 above)
Comment (knyazer) [reply test/test_viz.py:48]: Yeah, this does make sense, but also introduces a lot of internal dependecies: imagine we change collision resolution 'a little bit', and, while simulation is still physically plausible, your tests started failing.

I think we should not test the visualizer in any meaningful way for now: while it makes sense to test it, the complexity of tests would significantly exceed the complexity of the visualizer itself, which means it is not worth the effort.

File: test/test_viz.py (hunk 2 above)
Comment (knyazer) [reply test/test_viz.py:36]: This is not exactly visualization related: color is in fact a property of a body. While I cannot think about any other ways to use it, the rule of thumb is that if something is a property of a particular type of objects, you should include this something (color in this case) as property of the class. This is just sensible: while we indeed are going to perceive a performance hit, in practice it is firstly really small, and, secondly, there is a way to avoid it completely by marking the color as static: then this property won't be passed around while doing any operations with objects, which means 0 performance degradation.

I think that making a color of a body an ```Optional[Color]``` or just ```Color``` is a good enough decision. If you don't believe me, well, alright, you can open a discussion about it in the discussions tabs of the repo. 

I am ok with doing it your way if you can give a sensible explanation why is introducing a hierarchy of drawing classes is better than using ```Color``` as a property. For now I don't see why you would do this.

File: test/test_viz.py (hunk 3 above)
Comment (ivrolan) [reply test/test_viz.py:48]: I understand. However, it would be cool to have something like a "visualizer demo"  to try out how the drawing for different objects work. Should we include this in another folder `demos`?

Comment (knyazer) [reply test/test_viz.py:48]: Nope, end users don't care about how cool our visualizer is, they care that it works.

But I like the idea of having a folder like "examples", though its still not really what you mean. So id suggest that we don't do that until implementation of environments is ready

File: test/test_viz.py (hunk 2 above)
Comment (knyazer) [reply test/test_viz.py:36]: Hm, your point about showing something different when complicated situations occur does make sense. I guess I was thinking that users could use cotix to make any environment they would like. 

There is a nice abstraction in Jax that is usually used to do that, but I already messed it up a bit, so doing it your way is quite a decent solution.

To sum up: do it in any way you'd like. I stopped having a strong opinion on the topic.

File: test/test_viz.py (hunk 3 above)
Comment (knyazer) [reply test/test_viz.py:48]: I mean, look, more (good written) code is never worse than less. But implementing demos looks useless to me. Why would somebody who is trying to get his PPO, or whatever RL thing, working, care that visualizer can show collisions? They care only about the fact that it looks like the agent is playing well. I mean, generally, I think the standard way of determining that your model works is judging by metrics, and in our case we need a visualization only because we don't know what reward quantifies sufficiently good performance: so visualizer allow us to see if the agent can perform in alignment with our expectations of what good behaviour is, sort of a way to do human feedback. And I don't think that information about these collisions/balls catching robots is critical for judging about the overall performance of the agent. Yes, this informatino is useful for debugging, and overall for understanding why agent acts the way it does, but it is not critical, at least on the current stage.

Again, feel free to implement any of the two approaches, since both of them are sufficiently good to start with. I don't have a preference between the two anymore.

**Comment by knyazer** [comment]
Closing since completed with #52
